### ISSUE TITLE
Server crashing after every download ERR_STREAM_WRITE_AFTER_END

### ISSUE THREAD

**Issue by EmTeeAge** [Tzahi12345/YoutubeDL-Material#176]
Hi guys! Sorry for bothering you, but I recently tried to install Youtube-dl Material on my local Linux server (Debian 10), however, I've ran into some issues.

I followed the steps on the GitHub page and I tried both downloading the latest release and installing it the normal way and building it from source to see whether that could cause the issue. I also tried different directories on my server, running as root, etc. to eliminate folder permission errors. I'm also not using any reverse proxy right now to eliminate any other possible causes.

The issue is following:
Once I start the server using npm start and try to download a Youtube link (both video or audio), the server crashes with the following message:

```
> backend@1.0.0 start /var/www/youtubedl-material
> nodemon -q app.js

2020-07-22T23:07:42.658Z INFO: YoutubeDL-Material v4.1 started on PORT 17442
events.js:291
      throw er; // Unhandled 'error' event
      ^

Error [ERR_STREAM_WRITE_AFTER_END]: write after end
    at writeAfterEnd (_http_outgoing.js:646:15)
    at ServerResponse.end (_http_outgoing.js:766:7)
    at ServerResponse.end (/var/www/youtubedl-material/node_modules/compression/index.js:107:21)
    at /var/www/youtubedl-material/app.js:1947:9
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Emitted 'error' event on ServerResponse instance at:
    at writeAfterEndNT (_http_outgoing.js:705:7)
    at processTicksAndRejections (internal/process/task_queues.js:81:21) {
  code: 'ERR_STREAM_WRITE_AFTER_END'
}
```

This causes the website to break and become unusable. However, I found out that if I close the server using Ctrl + C in the terminal and start it again with npm start and then reload the webpage, it starts working again and plays the video/audio and allows me to download it, meaning that the download actually happened and once I go back to the main page, it remembers the download in the bottom section. Unfortunately upon downloading another video, the server crashes the same way and the process repeats.

This is from a Firefox console:
![image](https://user-images.githubusercontent.com/67438185/88238753-0e829b80-cc72-11ea-9031-7302ac02f466.png)

Is there anything I could try to solve the issue or any other info I can provide to help?

Thanks

**Comment by Tzahi12345** [comment]
Not a bother! That's what this site is for :)

Few questions:

1. Is it just the website that crashes, or the entire server becomes unresponsive? (If you reload, is it still broken?)

2. Do you get redirected to the video/audio player? Does the video/audio start playing at all?

3. Can you try using another browser, such as Chrome/Chromium?

-----

**Potential fix**

Looking at where the crash happened, it's at the end of the [/api/tomp4 route](https://github.com/Tzahi12345/YoutubeDL-Material/blob/v4.1/backend/app.js#L1947), specifically this on line 1947:

`res.end("yes");`

If you're comfortable going to the `backend/app.js` file, can you try commenting out that line like so:

`// res.end("yes");`

If this fixes the problem, then we've found the culprit, and I can go ahead and push a hotfix. Let me know how it goes!

**Comment by EmTeeAge** [comment]
Thanks for a quick response!

1. After it throws the error in a terminal, the website goes blank, but there's still a blue header at the top of the page with the title "YoutubeDL-Material", but the rest of the page is without any content. The URL is `http://192.168.1.195:17442/#/player;type=audio;uid=81c7fd03-734c-47af-9f27-1452a991b2b9`. If I reload now without restarting the server, Firefox will throw "Unable to connect" error.

2. However, if I restart the server first (just Ctrl + C in terminal and then `npm start` again) and then reload the webpage, I get a working page that plays both audio and video, while also allowing me to save the file. The rest works fine until I try to download another link from Youtube.

3. Yes, I tried ungoogled Chromium and Edge, but the issue is present across browsers.

So after commenting out what you suggested in the app.js file, it **still gave me the error**. When I looked at the lines in the code the error was pointing to, it was another occurance of res.end("yes") on the line 1916. Then I tried **commenting out all occurrences of res.end("yes")** in the app.js file and that **fixed the issue.**

But then I tried deleting an entry in the "Your audio files are here" element and it threw a different error. 
I wanted to test out the errors that it's throwing when deleting things from the "Your audio/video files are here", so I set up 4 case scenarios:

- Delete audio
- Delete and blacklist audio
- Delete video
- Delete and blacklist video

1. Error when you click "Delete" in the audio section:

```
2020-07-23T01:20:53.498Z INFO: YoutubeDL-Material v4.1 started on PORT 17442
(node:23423) UnhandledPromiseRejectionWarning: ReferenceError: filePath is not defined
    at /var/www/youtubedl-material/app.js:890:39
    at new Promise (<anonymous>)
    at deleteAudioFile (/var/www/youtubedl-material/app.js:885:12)
    at /var/www/youtubedl-material/app.js:2476:9
    at Layer.handle [as handle_request] (/var/www/youtubedl-material/node_modules/express/lib/router/layer.js:95:5)
    at next (/var/www/youtubedl-material/node_modules/express/lib/router/route.js:137:13)
    at optionalJwt (/var/www/youtubedl-material/app.js:1841:12)
    at Layer.handle [as handle_request] (/var/www/youtubedl-material/node_modules/express/lib/router/layer.js:95:5)
    at next (/var/www/youtubedl-material/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/var/www/youtubedl-material/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/var/www/youtubedl-material/node_modules/express/lib/router/layer.js:95:5)
    at /var/www/youtubedl-material/node_modules/express/lib/router/index.js:281:22
    at Function.process_params (/var/www/youtubedl-material/node_modules/express/lib/router/index.js:335:12)
    at next (/var/www/youtubedl-material/node_modules/express/lib/router/index.js:275:10)
    at compression (/var/www/youtubedl-material/node_modules/compression/index.js:220:5)
    at Layer.handle [as handle_request] (/var/www/youtubedl-material/node_modules/express/lib/router/layer.js:95:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:23423) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:23423) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

2. Error when you click "Delete and blacklist" in the audio section:

`Same error as in Case 1.`

3. Error when you click "Delete" in the video section:

`2020-07-23T01:31:22.084Z ERROR: Archive could not be found.`

4. Error when you click "Delete and blacklist" in the audio section:

`2020-07-23T01:31:22.084Z ERROR: Archive could not be found.`

So the only issue now is when deleting audio files, but I have no idea if or how it's related to commenting out the string that you mentioned.

Also I'm a bit worried that it's something on my side with dependencies that could be causing the issue, so I don't want you to break your code that works for others fine :)

Thanks

**Comment by Tzahi12345** [comment]
>So after commenting out what you suggested in the app.js file, it still gave me the error. When I looked at the lines in the code the error was pointing to, it was another occurance of res.end("yes") on the line 1916. Then I tried commenting out all occurrences of res.end("yes") in the app.js file and that fixed the issue.

Awesome! I've gone ahead and pushed a fix for this, [see commit here](https://github.com/Tzahi12345/YoutubeDL-Material/commit/c800308b9dda74da606b0bdf6a8fa14250c14afb). `res.end("yes")` is a redundant call, it didn't need to be there and so it's safe to remove.

>So the only issue now is when deleting audio files, but I have no idea if or how it's related to commenting out the string that you mentioned.

Turns out it's not, you found another bug! Not sure how I didn't find this before (it's been there for a few weeks now!), I've gone ahead and patched it, [see commit here](https://github.com/Tzahi12345/YoutubeDL-Material/commit/c1c57135bae7951338294f0ab4cc6624d5a08066).

In terms of the error finding the archive, did you download the videos before you enabled archive mode? This is the most likely cause for this error.

>Also I'm a bit worried that it's something on my side with dependencies that could be causing the issue, so I don't want you to break your code that works for others fine :)

I'm not sure why the `res.send("yes")` gave you problems, I've never seen it. Maybe it's related to your node.js version (I use v12).

Anyways, the fixes I pushed are 100% safe, and the second bug you found affects all users, what a great find. You can run `git pull` to get the latest updates, no re-build should be required as the changes are all on the backend. Or you can do it the easy way and get on the `nightly` tag on Docker, but that's only if you're O.K. with using Docker.

Let me know if the changes fix the bugs! Awesome job debugging, saved me a lot of investigating :)

**Comment by EmTeeAge** [comment]
That's awesome!

Sorry for the delay, so I tried replacing the app.js file with the chages you've made and as far as I know, it works perfectly! Just to be sure, I installed the whole thing again starting with the new app.js file and I haven't noticed any issues, everything seems to be working as intended.

You've mentioned that you're running node.js v12, so I checked and I'm using v14.5.0, which might have been the cause.

Just a quick question about the archive - I enabled archive in default.json before starting the server, downloaded a file and when I deleted it, the deletion was successful, however I still got the "Archive could not be found" error in the terminal. Doesn't seem to effect anything other that throwing the error.

Anyways, thanks a lot! I'm glad I could help :)
